src/main/scala/is/hail/distributedmatrix/Keys.scala

```diff
@@ -0,0 +1,83 @@
+package is.hail.distributedmatrix
+
+import is.hail.annotations.Annotation
+import is.hail.expr.{JSONAnnotationImpex, Parser}
+import is.hail.expr.types.Type
+import is.hail.utils._
+import org.apache.spark.SparkContext
+import org.json4s.jackson.Serialization
+import org.json4s.{JArray, JValue}
+import org.json4s.jackson.JsonMethods.parse
+
+case class KeysImpex(typeStr: String, valuesJ: JValue)
+
+object Keys {
+  def unify(left: Option[Keys], right: Option[Keys], msg: String): Option[Keys] = {
+    (left, right) match {
+      case (Some(l), Some(r)) =>
+        l.assertSame(r, msg)
+        left
+      case (Some(_), _) => left
+      case _ => right
+    }
+  }
+  
+  def read(sc: SparkContext, uri: String): Keys = {
+    val KeysImpex(typeStr, JArray(arr)) =  sc.hadoopConfiguration.readTextFile(uri)(in =>
+      try {
+        val json = parse(in)
+        json.extract[KeysImpex]
+      } catch {
+        case e: Exception => fatal(
+          s"""corrupt or outdated matrix key data.
+             |  Recreate with current version of Hail.
+             |  Detailed exception:
+             |  ${ e.getMessage }""".stripMargin)
+      })
+    
+    val typ = Parser.parseType(typeStr)
+    val values = arr.map(JSONAnnotationImpex.importAnnotation(_, typ)).toArray
+    
+    new Keys(typ, values)
+  }
+}
+
+class Keys(val typ: Type, val values: Array[Annotation]) {
+  def length: Int = values.length
+  
+  def assertSame(that: Keys, msg: String = "") {
+    if (typ != that.typ)
+      fatal(msg + "Keys have different types: " +
+        s"${typ.toPrettyString(compact = true)}, ${ that.typ.toPrettyString(compact = true) }")
+
+    if (values.length != that.values.length)
+      fatal(msg + s"Differing number of keys: $length, ${ that.length }")
+
+    var i = 0
+    while (i < length) {
+      if (values(i) != that.values(i))
+        fatal(msg + s"Key mismatch at index $i: ${typ.str(values(i))}, ${typ.str(that.values(i))}")
+      i += 1
+    }
+  }
+  
+  def filter(pred: Annotation => Boolean): Keys = new Keys(typ, values.filter(pred))
+  
+  def filter(keep: Array[Int], check: Boolean = true): Keys = {
+    if (check)
```

Comment:
Why?

Reply:
To prevent this requirement from being checked twice in KeyedBlockMatrix filter, since the same check is done on the level of the BlockMatrix filter. I've now removed the added complexity. Another option would be to turn the check into a comment, since this function is currently only used in situations where the check is redundant. Making an additional class for ordered indices seems like overkill at this stage.